File: python/64_Minimum_Path_Sum.py
#!/usr/bin/python

class Solution(object):
    def minPathSum(self, grid):
    	m = len(grid)
    	n = len(grid[0])
        dp = [[0] * n for i in range(m)]
        dp[0][0] = grid[0][0]
        for i in range(m - 1):
        	dp[i + 1][0] = dp[i][0] + grid[i + 1][0]
        for j in range(n - 1):
        	dp[0][j + 1] = dp[0][j] + grid[0][j + 1]
        for i in range(m - 1):
        	for j in range(n - 1):
        		dp[i + 1][j + 1] = min(dp[i + 1][j], dp[i][j + 1]) + grid[i + 1][j + 1]
        return dp[m - 1][n - 1]

def main():
    s = Solution()
    grid = [[1,2,3], [4, 2, 1]]

    print s.minPathSum(grid)

if __name__ == "__main__":
    main()

File: c++/64_Minimum_Path_Sum.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int> >& grid) {
    	int m = grid.size();
    	int n = grid[0].size();
    	int dp[m][n] = {{0}};
    	if (m == 1 && n == 1)
    		return grid[0][0];
    	dp[0][0] = grid[0][0];
    	for (int i = 0; i < m - 1; i++)
    		dp[i + 1][0] = dp[i][0] + grid[i + 1][0];
    	for (int j = 0; j < n - 1; j++)
    		dp[0][j + 1] = dp[0][j] + grid[0][j + 1];
    	for (int i = 0; i < m - 1; i++)
	    	for (int j = 0; j < n - 1; j++) 
	    		dp[i + 1][j + 1] = min(dp[i + 1][j], dp[i][j + 1]) + grid[i + 1][j + 1];
	    return dp[m - 1][n - 1];
    }
};


int main() {
	Solution s;


}